content/blog/2020-12-20-loop-flatten/index.md

```diff
@@ -0,0 +1,199 @@
+    2|       |#include <stdlib.h>
+    3|       |
+    4|      2|#define N 11
+    5|    441|#define M 20
+    6|       |
+    7|      1|int main() {
+    8|      1|  int init_val = 30;
+    9|      1|  int res[M*N];
+   10|      1|  LOOP:
+   11|      2|  for (int i = 0; i < N; i++) {
+   12|    221|    for (int j = 0; j < M; j++) {
+   13|    220|      res[i*M+j] = init_val;
+   14|    220|    }
+   15|      1|  }
+   16|      1|  printf("Multiplication res: %d; \n", res[0]);
+   17|      1|  return 0;
+   18|      1|}
+```
+The first line is the sample result I printed out to verify the correctness of the optimized code. And then the following code is the optimized one with the updated inner loop bound. As you can see from line 11, the outer loop was being executed only 1 time. The number is 2 here means the second time, it checks the loop bound and exits the outer loop. And accordingly, the inner loop is being executed `N*M` times, which is 220 from line 13. This naïve pass is being tested with other small benchmarks I created with perfect nested loop and passed all the tests.
+
+2. Performance gain<br/>
+As we mentioned before, the loop flattening pass itself does not offer any benefits for performance. But this technique helps to better explore the parallelism when combining with the loop unrolling technique. Specifically, we can apply larger unroll parameters to further explore the parallelism across the loop iterations. Also, this technique is used to support other optimizing pragmas (like loop pipelining in HLS tools) for better performance.
+
+
+## Conclusion
+In summary, loop flattening is a useful optimizing technique for supporting loop unrolling and other HLS related optimizations. LLVM is an awesome infrastructure that provides us powerful APIs for creating our own pass. Finally, I would like to mention that the loop flattening pass has just been merged into the official LLVM github repository (LLVM v12.0.0). Here is the [source code](https://llvm.org/doxygen/LoopFlatten_8cpp_source.html). Unlike my naive implementation, this pass can actually flatten the loop as shown in the following example:
+
+```javascript
+ // from nested loop:
+ for (int i = 0; i < N; ++i)
+   for (int j = 0; j < M; ++j)
+     f(A[i*M+j]);
+ // into one loop:
+ for (int i = 0; i < (N*M); ++i)
+   f(A[i]);
+```
+
+However, it also has some constraints which are elaborated in the [source code](https://llvm.org/doxygen/LoopFlatten_8cpp_source.html).
+
+## References
+[1] The LLVM Compile Infrastructure: https://llvm.org/ <br/>
+[2] LLVM user reference manual: https://llvm.org/docs/LangRef.html<br/>
+[3] LLVM developer reference manual: http://llvm.org/doxygen/<br/>
+[4] LLVM official github repository: https://github.com/llvm/llvm-project<br/>
+[5] Loop flatten techniques applied in software pipelining (converting loop level parallelism into instruction parallelism): https://people.ece.uw.edu/hauck/publications/LoopFlattening.pdf<br/>

```

Comment:
I generally think it's probably better to link to stuff in context, when you're writing for the Web. Unlike when you're writing a PDF, it's just as easy to do that and more convenient to read than having references at the end.